command/server/tls_util.go

```diff
@@ -0,0 +1,178 @@
+package server
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"net"
+	"os"
+	"strings"
+	"time"
+)
+
+type CaCert struct {
+	PEM      string
+	Template *x509.Certificate
+	Signer   crypto.Signer
+}
+
+// GenerateCert creates a new leaf cert from provided CA template and signer
+func GenerateCert(caCertTemplate *x509.Certificate, caSigner crypto.Signer) (string, string, error) {
+	// Create the private key
+	signer, keyPEM, err := privateKey()
+	if err != nil {
+		return "", "", err
+	}
+
+	// The serial number for the cert
+	sn, err := serialNumber()
+	if err != nil {
+		return "", "", err
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return "", "", err
+	}
+
+	if hostname == "" {
+		hostname = "localhost"
+	}
+
+	// Create the leaf cert
+	template := x509.Certificate{
+		SerialNumber:          sn,
+		Subject:               pkix.Name{CommonName: hostname},
+		BasicConstraintsValid: true,
+		KeyUsage:              x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
+		NotAfter:              time.Now().Add(365 * 24 * time.Hour),
+		NotBefore:             time.Now().Add(-1 * time.Minute),
+		IPAddresses:           []net.IP{net.ParseIP("127.0.0.1")},
+		DNSNames:              []string{"localhost", "localhost4", "localhost6", "localhost.localdomain"},
+	}
+
+	bs, err := x509.CreateCertificate(
+		rand.Reader, &template, caCertTemplate, signer.Public(), caSigner)
+	if err != nil {
+		return "", "", err
+	}
+	var buf bytes.Buffer
+	err = pem.Encode(&buf, &pem.Block{Type: "CERTIFICATE", Bytes: bs})
+	if err != nil {
+		return "", "", err
+	}
+
+	return buf.String(), keyPEM, nil
+}
+
+// GenerateCA generates a new self-signed CA cert and returns a
+// CaCert struct containing the PEM encoded cert,
+// X509 Certificate Template, and crypto.Signer
+func GenerateCA() (*CaCert, error) {
+	// Create the private key we'll use for this CA cert.
+	signer, _, err := privateKey()
+	if err != nil {
+		return nil, err
+	}
+
+	// The serial number for the cert
+	sn, err := serialNumber()
+	if err != nil {
+		return nil, err
+	}
+
+	signerKeyId, err := keyId(signer.Public())
```

Comment:
You could replace this with a call to `sdk/helper/certutil`'s `GetSubjKeyID` which _is_ RFC compliant and means you don't need the implementation here. ;-)

Reply:
This let me remove a function, nice!